File: python/leetcode/binary-tree-paths.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def binaryTreePaths(self, root):
        """
        :type root: TreeNode
        :rtype: List[str]
        """
        this_path = ""
        paths = []
        if root is None:
            return paths
        return self.findPath(root, this_path, paths)
    
    def findPath(self, root, this_path, paths):
        if root.left is None and root.right is None:
            this_path = "%s%s" % (this_path, root.val)
            paths.append(this_path)
        if root.left is not None:
            paths = self.findPath(root.left, "%s%s->" % (this_path, root.val), paths)
        if root.right is not None:
            paths = self.findPath(root.right, "%s%s->" % (this_path, root.val), paths)
        return paths

File: java/leetcode/binary-tree-paths.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {

    public List<String> binaryTreePaths(TreeNode root) {
        String thisPath = "";
        List<String> paths = new ArrayList();
        if (root == null) { return paths; }
        findPath(root, thisPath, paths);
        return paths; 
    }
    
    private void findPath(TreeNode root, String thisPath, List<String> paths) {
        if (root.left == null && root.right == null) {
            thisPath += root.val;
            paths.add(thisPath);
        }
        if (root.left != null) {
            findPath(root.left, thisPath + root.val + "->", paths);
        }
        if (root.right != null) {
            findPath(root.right, thisPath + root.val + "->", paths);
        }
    }
}
